File: HOT100/HOT100p1/6.矩阵/73.矩阵置零.cpp
/*
 * @lc app=leetcode.cn id=73 lang=cpp
 *
 * [73] 矩阵置零
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        bool _m0flag = 0, _n0flag = 0;
        int m = matrix.size(), n = matrix[0].size();
        for (int i = 0; i < m; i ++ )
        {
            if (matrix[i][0] == 0) 
            {
                _m0flag = 1;
                break;
            }
        }
        for (int j = 0; j < n; j ++ )
        {
            if (matrix[0][j] == 0)
            {
                _n0flag = 1;
                break;
            }
        }
        for (int i = 1; i < m; i ++ )
        {
            for (int j = 1; j < n; j ++ )
            {
                if (matrix[i][j] == 0)
                {
                    matrix[0][j] = matrix[i][0] = 0;
                }
            }
        }
        for (int i = 1; i < m; i ++ )
        {
            for (int j = 1; j < n; j ++ )
            {
                if (matrix[i][0] == 0 || matrix[0][j] == 0)
                {
                    matrix[i][j] = 0;
                }
            }
        }
        if (_m0flag)
        {
            for (int i = 0; i < m; i ++ )
            {
                matrix[i][0] = 0;
            }
        }
        if (_n0flag)
        {
            for (int j = 0; j < n; j ++ )
            {
                matrix[0][j] = 0;
            }
        }
    }
};
// @lc code=end

File: HOT100/HOT100p1/6.矩阵/73.矩阵置零.py
#
# @lc app=leetcode.cn id=73 lang=python3
#
# [73] 矩阵置零
#
from typing import List
# @lc code=start
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0])
        flagl, flagh = False, False
        for i in range(m):
            if matrix[i][0] == 0:
                flagl = True
                break
        for j in range(n):
            if matrix[0][j] == 0:
                flagh = True
                break
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        for i in range(1, m):
            if matrix[i][0] == 0:
                matrix[i] = [0] * n
        for j in range(1, n):
            if matrix[0][j] == 0:
                for u in range(m):
                    matrix[u][j] = 0
        if flagl:
            for u in range(m):
                matrix[u][0] = 0
        if flagh:
            matrix[0] = [0] * n


# @lc code=end

